Add check version functionality

bnkai opened the pull request:
Adds initial support for new version checking. Mentioned also in #288 
As it is now it uses the github api to get the shorthash commits from the repo tags.
It matches the current version ( `v0.0.0-alpha` or 'v0.0.0-alpha_dev' ) to the respective repo tags and gets the related commit hash from there.If it's different than the local that means that a new release build is available. 
It first prints a message in the console when the app is started and you can see it also when you open the about tab from the settings.An extra button is also available for manual checking.

for dev testing the build , since version is added through TRAVIS you need to add it to the ldflags in the Makefile
`-X 'github.com/stashapp/stash/pkg/api.version=v0.0.0-alpha_dev' `
or
`-X 'github.com/stashapp/stash/pkg/api.version=v0.0.0-alpha'`


![console](https://user-images.githubusercontent.com/48220860/71551775-b3e4d200-29f7-11ea-9298-b1a2d2ba0e42.png)

![ui](https://user-images.githubusercontent.com/48220860/71551777-c65f0b80-29f7-11ea-9782-6c546a21f984.png)

Leopere commented:
Eventually, I really would like to update automagically by default.  Potentially allow people to disable the update mechanism via some command line launch argument or something not in the UI.

WithoutPants commented:
Reviewed but not yet tested.

WithoutPants commented on pkg/api/check_version.go:
```diff
@@ -0,0 +1,101 @@
+
+		data, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return "", fmt.Errorf("Github API read respone failed: %s", err)
```
Typo - respon**s**e

bnkai replied:
fiixe

WithoutPants commented on pkg/api/check_version.go:
```diff
@@ -0,0 +1,101 @@
+
+	branch, _, _ := GetVersion()
+	if branch == "" {
+		return "", fmt.Errorf("Stash doesn't have a version.Version check not supported.")
```
Add space between sentences.

bnkai replied:
fixed

WithoutPants commented on Makefile:
```diff
@@ -6,7 +6,7 @@ endif
 	$(eval DATE := $(shell go run scripts/getDate.go))
 	$(eval GITHASH := $(shell git rev-parse --short HEAD))
-	$(SET) CGO_ENABLED=1 $(SEPARATOR) go build -mod=vendor -v -ldflags "-X 'github.com/stashapp/stash/pkg/api.buildstamp=$(DATE)' -X 'github.com/stashapp/stash/pkg/api.githash=$(GITHASH)'"
+	$(SET) CGO_ENABLED=1 $(SEPARATOR) go build -mod=vendor -v -ldflags "-X 'github.com/stashapp/stash/pkg/api.buildstamp=$(DATE)' -X 'github.com/stashapp/stash/pkg/api.githash=$(GITHASH)' "
```
Why is this change necessary?

bnkai replied:
Sorry  , missed that. It was leftover from manual editing the Make file to add the version and so a space was left there.

WithoutPants commented:
Since we are getting the URL anyway, is there any reason we can't or shouldn't include a direct link to the latest download in the about page?

bnkai commented:
> Since we are getting the URL anyway, is there any reason we can't or shouldn't include a direct link to the latest download in the about page?

The API endpoint i am quering is this one [tags](https://api.github.com/repos/stashapp/stash/tags).
It only has the versions and the URL of the zipfile , tarball **source code** no releases.
There is also the [releases](https://api.github.com/repos/stashapp/stash/releases) one that can be queried ,  it is more complex to parse though.
I'll have a look at it , switching to WIP till i settle this.

ueaslsef commented:
The download path should be rather straight forward: `https://github.com/stashapp/stash/releases/download/$version/stash-$platform`. Maybe we could cheat and skip the releases-api parsing all together and just use this until we found a better way? It could even be a static string inserted by the ci on build time.

bnkai commented:
@ueaslsef  since we are doing it we might as well go all the way  ..
The releases API has extra stuff also like times downloaded or date released. Do we want any of that ?

bnkai commented:
squashed and rebased , also fixed a merge conflict due to latest commits.
this the latest version with the proper release added as download link.
![releases](https://user-images.githubusercontent.com/48220860/71635204-225da600-2c2b-11ea-9cf3-967885a64042.png)

WithoutPants commented on pkg/api/check_version.go:
```diff
@@ -0,0 +1,179 @@
+	_, githash, _ = GetVersion()
+	latest, _, err := GetLatestVersion(true)
+	if err != nil {
+		logger.Errorf("Couldn't find latest version :%s\n", err)
```
Change to `version: %s`. Looks weird in the log currently. The `\n` is unnecessary and pollutes the output.

bnkai replied:
i think i added it for the console, removed anyway

WithoutPants replied:
I need to proofread my feedback. This was supposed to be two separate items: the colon position and the `\n` character. Currently, the log looks like this:
```
time="2020-01-05T09:54:31+11:00" level=error msg="Couldn't find latest version :Stash doesn't have a version. Version check not supported."
```

Should be `version: %s` rather than `version :%s`.

bnkai replied:
missed that, fixed.

WithoutPants commented on ui/v2/src/components/Settings/SettingsAboutPanel.tsx:
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
           <tr>
-            <td>Build time:</td>
-            <td>{data.version.build_time}</td>
+            <td><Button minimal={true} onClick={() => refetch()} text="Check for new version!" /></td>
```
Having `minimal={true}` makes the button look weird in this context when not moused over. I'd get rid of it. I'm not sure that the exclamation mark is necessary, but that's just a subjective aesthetic opinion.

bnkai replied:
i tried the combo minimal and ! and plain button without the !
i'll change back to plain button

WithoutPants commented on pkg/api/check_version.go:
```diff
@@ -0,0 +1,179 @@
+		logger.Errorf("Couldn't find latest version :%s\n", err)
+	} else {
+		if githash == latest {
+			logger.Infof("Version: (%s) is already the latest released.\n", latest)
```
Remove `\n`

WithoutPants commented on pkg/api/check_version.go:
```diff
@@ -0,0 +1,179 @@
+		if githash == latest {
+			logger.Infof("Version: (%s) is already the latest released.\n", latest)
+		} else {
+			logger.Infof("New version: (%s) available.\n", latest)
```
Remove `\n`.

WithoutPants commented on ui/v2/src/components/Settings/SettingsAboutPanel.tsx:
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
-            <td>Build hash:</td>
-            <td>{data.version.hash}</td>
+            < td>Latest Version Build Hash: </td>
+            < td>{maybeRenderLatestVersion()} </td>
```
Spaces before the `td` tag need to be fixed.

WithoutPants commented on ui/v2/src/components/Settings/SettingsAboutPanel.tsx:
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
+    );
+  }
+
+  function RenderLatestVersion() {
```
First letter should be lowercase.

WithoutPants commented on ui/v2/src/components/Settings/SettingsAboutPanel.tsx:
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
+      {!!er ? <span>{er.message}</span> : undefined}
       {renderVersion()}
+      {!dt || ld || networkStatus == 4 ? <Spinner size={Spinner.SIZE_SMALL} /> : undefined}
+      {RenderLatestVersion()}
```
I think these variables (`dt`, `ld`, `er`) need to be better named.

I would make it so that if the spinner is showing (ie the query is in progress) then `renderLatestVersion` is not called.

bnkai replied:
should be covered with latest commit ....
